File: src/lib/link-resolver.js
/**
 * Global link resolver — maps every external URL pattern we know about to its
 * corresponding corpus storage key, and packages the rules into a callback
 * shape compatible with `htmlToMarkdown`'s `linkResolver` opt and the audit
 * command's classifier.
 *
 * URL→key rules live in one place so adding a new source means adding one
 * pattern instead of editing per-adapter resolvers.
 */

/** Pure mapping rules: URL → candidate corpus key, no DB access. */
const RULES = [
  // developer.apple.com/documentation/<framework>/<rest>  → <framework>/<rest>
  // The framework slug is Apple's lowercase-no-hyphen form (`swiftui`, `appkit`).
  // Strip trailing slashes; keep the rest of the path verbatim and lowercase.
  {
    test: (u) => u.hostname === 'developer.apple.com' && /^\/documentation\//.test(u.pathname),
    map: (u) => {
      const path = u.pathname.replace(/^\/documentation\//, '').replace(/\/+$/, '')
      return path ? path.toLowerCase() : null
    },
  },

  // developer.apple.com/design/human-interface-guidelines/<rest>
  //   → design/human-interface-guidelines/<rest>
  {
    test: (u) => u.hostname === 'developer.apple.com' && u.pathname.startsWith('/design/'),
    map: (u) => {
      const path = u.pathname.replace(/^\//, '').replace(/\/+$/, '')
      return path ? path.toLowerCase() : null
    },
  },

  // developer.apple.com/library/archive/<archive-path>  → apple-archive/<archive-path>
  // Archive keys preserve original case ("AppleApplications/Conceptual/...").
  {
    test: (u) => u.hostname === 'developer.apple.com' && u.pathname.startsWith('/library/archive/'),
    map: (u) => {
      const path = u.pathname.replace(/^\/library\/archive\//, '').replace(/\/+$/, '')
      if (!path) return null
      // The catalog applies an index/match collapse to terminal HTML files
      // ("…/Foo.html" with parent "Foo" maps to "…/Foo"). Mirror that here so
      // the candidate key matches the stored format.
      const match = path.match(/^(.*)\/([^/]+)\.(html|htm)$/i)
      if (match) {
        const [, dir, base, ext] = match
        const parent = dir.split('/').pop() ?? ''
        if (base.toLowerCase() === 'index' || base.toLowerCase() === parent.toLowerCase()) {
          return `apple-archive/${dir}`
        }
        return `apple-archive/${dir}/${base}.${ext.toLowerCase()}`
      }
      return `apple-archive/${path}`
    },
  },

  // developer.apple.com/videos/play/wwdc<year>/<id>/  → wwdc/wwdc<year>-<id>
  {
    test: (u) => u.hostname === 'developer.apple.com' && /^\/videos\/play\/wwdc\d{4}\/\d+\/?$/.test(u.pathname),
    map: (u) => {
      const m = u.pathname.match(/^\/videos\/play\/(wwdc\d{4})\/(\d+)/)
      return m ? `wwdc/${m[1]}-${m[2]}` : null
    },
  },

  // docs.swift.org/swift-book/documentation/the-swift-programming-language/<chapter>
  //   → swift-book/<chapter>
  // The corpus key uses CamelCase chapter names ("LanguageGuide/TheBasics");
  // the URL uses lowercase ("languageguide/thebasics"). Without a chapter
  // index we can only return the lowercased rest and let the caller's
  // knownKeys.has() pass it through after lower-cased comparison.
  {
    test: (u) => u.hostname === 'docs.swift.org' && u.pathname.startsWith('/swift-book/'),
    map: (u) => {
      const rest = u.pathname.replace(/^\/swift-book\//, '').replace(/\/+$/, '')
      return rest ? `swift-book/${rest}` : 'swift-book'
    },
  },

  // docs.swift.org/compiler/<rest>  → swift-compiler/<rest>
  {
    test: (u) => u.hostname === 'docs.swift.org' && u.pathname.startsWith('/compiler/'),
    map: (u) => {
      const rest = u.pathname.replace(/^\/compiler\//, '').replace(/\/+$/, '')
      return rest ? `swift-compiler/${rest}` : 'swift-compiler'
    },
  },

  // docs.swift.org/swiftpm/<rest>  → swift-package-manager/<rest>
  {
    test: (u) => u.hostname === 'docs.swift.org' && u.pathname.startsWith('/swiftpm/'),
    map: (u) => {
      const rest = u.pathname.replace(/^\/swiftpm\//, '').replace(/\/+$/, '')
      return rest ? `swift-package-manager/${rest}` : 'swift-package-manager'
    },
  },

  // swift.org/migration/<rest>  → swift-migration-guide/<rest>
  {
    test: (u) => isSwiftOrg(u) && u.pathname.startsWith('/migration/'),
    map: (u) => {
      const rest = u.pathname.replace(/^\/migration\//, '').replace(/\/+$/, '')
      return rest ? `swift-migration-guide/${rest}` : 'swift-migration-guide'
    },
  },

  // swift.org/swift-evolution/proposals/<NNNN-name>.html  → swift-evolution/<NNNN-name>
  // Some proposals are also linked as github.com/apple/swift-evolution/blob/main/proposals/...
  {
    test: (u) => isSwiftOrg(u) && /^\/swift-evolution\/proposals\/\d{4}-/.test(u.pathname),
    map: (u) => {
      const m = u.pathname.match(/\/proposals\/(\d{4}-[^/]+?)(?:\.md|\.html)?\/?$/)
      return m ? `swift-evolution/${m[1]}` : null
    },
  },
  {
    test: (u) => u.hostname === 'github.com' && /^\/(?:apple|swiftlang)\/swift-evolution\/(?:blob|tree)\//.test(u.pathname),
    map: (u) => {
      const m = u.pathname.match(/\/proposals\/(\d{4}-[^/]+?)(?:\.md|\.html)?\/?$/)
      return m ? `swift-evolution/${m[1]}` : null
    },
  },

  // NB: there is intentionally no generic `swift.org/<anything> → swift-org/`
  // rule. The set of swift-org pages we publish is curated; mapping every
  // swift.org URL would internalize blog posts, marketing pages, and
  // unrelated docs. The swift-org adapter passes its CURATED_PATHS set to
  // `createLinkResolver` via `opts.swiftOrgPaths` so only listed paths are
  // claimed (see createLinkResolver implementation below).
]

const SWIFT_ORG_REDIRECTS = {
  'documentation/concurrency':       'swift-migration-guide/documentation/migrationguide',
  'documentation/package-manager':   'swift-package-manager/documentation/packagemanagerdocs',
  'documentation/tspl':              'swift-book/The-Swift-Programming-Language',
}

function isSwiftOrg(u) {
  return u.hostname === 'swift.org' || u.hostname === 'www.swift.org'
}

/**
 * Try every URL→key rule in order and return the first non-null match.
 * Pure (no I/O); safe to call on any URL string.
 *
 * @param {string} url
 * @returns {string|null} candidate corpus key
 */
export function mapUrlToKey(url) {
  if (typeof url !== 'string' || !url) return null
  let parsed
  try { parsed = new URL(url) } catch { return null }

  // Apply known swift.org redirect aliases before hostname rules so they
  // win over the generic swift-org fallback.
  if (isSwiftOrg(parsed)) {
    const path = parsed.pathname.replace(/^\/+/, '').replace(/\/+$/, '')
    if (path in SWIFT_ORG_REDIRECTS) return SWIFT_ORG_REDIRECTS[path]
  }

  for (const rule of RULES) {
    if (rule.test(parsed)) {
      const key = rule.map(parsed)
      if (key) return key
    }
  }
  return null
}

/**
 * Build a link-resolver callback compatible with `htmlToMarkdown`'s
 * `linkResolver` opt: returns the rewritten href string, or `null` to
 * unwrap the anchor (keep the inner text only). The original href is
 * returned when no rewrite applies.
 *
 * @param {object} opts
 * @param {Set<string>} [opts.knownKeys] Set of corpus keys that exist.
 *   When supplied, every pattern match is verified before internalization.
 *   Without it, the strict pattern-based rules are trusted.
 * @param {Set<string>} [opts.swiftOrgPaths] Curated swift.org paths. When
 *   present, swift.org URLs whose path-without-trailing-slash is in this
 *   set are internalized to `/docs/swift-org/<path>/`.
 * @param {string} [opts.sourceUrl] Absolute URL of the page being parsed,
 *   used to resolve relative links.
 * @param {string} [opts.docsBase] Internal route prefix (default `/docs`).
 *   The returned hrefs are `${docsBase}/<key>/`.
 * @returns {(href: string) => string}
 */
export function createLinkResolver(opts = {}) {
  const knownKeys = opts.knownKeys ?? null
  const swiftOrgPaths = opts.swiftOrgPaths ?? null
  const docsBase = opts.docsBase ?? '/docs'
  const baseUrl = opts.sourceUrl
  let base
  try { base = baseUrl ? new URL(baseUrl) : null } catch { base = null }

  return (rawHref) => {
    if (typeof rawHref !== 'string' || !rawHref) return rawHref
    // Bail on schemes we never want to rewrite.
    if (/^(?:mailto:|tel:|javascript:|data:|#)/i.test(rawHref)) return rawHref

    let abs
    try {
      abs = base ? new URL(rawHref, base) : new URL(rawHref)
    } catch {
      return rawHref
    }

    // 1. Already an internal /docs/<key>/ route — leave alone.
    if (abs.origin === (base?.origin ?? abs.origin) && abs.pathname.startsWith(`${docsBase}/`)) {
      return rawHref
    }

    // 2. Try to internalize against the structured-pattern rules. With
    //    `knownKeys` we verify the key exists; without it we trust every
    //    pattern match.
    const candidate = mapUrlToKey(abs.toString())
    if (candidate && (knownKeys === null || knownKeys.has(candidate))) {
      const fragment = abs.hash || ''
      return `${docsBase}/${candidate}/${fragment}`
    }

    // 3. swift.org generic-path opt-in: internalize only when the path is
    //    in the caller's curated-path set. Avoids capturing /blog, /jobs,
    //    /support, etc.
    if (swiftOrgPaths && isSwiftOrg(abs)) {
      const path = abs.pathname.replace(/^\/+/, '').replace(/\/+$/, '')
      const variants = [path, `${path}.html`]
      for (const v of variants) {
        if (swiftOrgPaths.has(v)) {
          const fragment = abs.hash || ''
          return `${docsBase}/swift-org/${v}/${fragment}`
        }
      }
    }

    // 4. Otherwise return the absolute URL — at least the link works as
    //    external content and doesn't dangle on our host.
    return abs.toString()
  }
}

/**
 * Classify a single URL for audit purposes.
 *
 * @param {string} url
 * @param {object} opts
 * @param {Set<string>} opts.knownKeys
 * @param {string} [opts.docsBase] internal route prefix (default `/docs`)
 * @returns {{ category: string, internalKey?: string, normalized?: string }}
 *   category ∈ {fragment, internal_ok, internal_broken,
 *               external_resolvable, external, relative_broken}
 */
export function classifyLink(url, opts) {
  const knownKeys = opts.knownKeys
  const docsBase = opts.docsBase ?? '/docs'

  if (typeof url !== 'string' || !url) {
    return { category: 'relative_broken' }
  }

  // Fragment-only links are always page-local.
  if (url.startsWith('#')) return { category: 'fragment' }

  // Non-http schemes are external by definition.
  if (/^(?:mailto:|tel:|javascript:|data:)/i.test(url)) {
    return { category: 'external', normalized: url }
  }

  // Internal /docs/<key>/ links — verify the key resolves.
  if (url.startsWith(`${docsBase}/`)) {
    // Strip trailing slash + fragment + querystring to get the key.
    const m = url.slice(docsBase.length + 1).match(/^([^?#]+?)\/?(?:[?#].*)?$/)
    const key = m ? decodeURIComponent(m[1]) : null
    if (key && knownKeys.has(key)) {
      return { category: 'internal_ok', internalKey: key }
    }
    return { category: 'internal_broken', internalKey: key ?? url }
  }

  // Any other relative path — flag as broken (it would resolve against the
  // current host, which is rarely what we want).
  if (url.startsWith('/')) {
    return { category: 'relative_broken', normalized: url }
  }

  // Absolute URL: try to internalize, otherwise mark external.
  let abs
  try { abs = new URL(url) } catch { return { category: 'relative_broken', normalized: url } }
  const candidate = mapUrlToKey(abs.toString())
  if (candidate && knownKeys.has(candidate)) {
    return { category: 'external_resolvable', internalKey: candidate, normalized: url }
  }
  return { category: 'external', normalized: url }
}

File: swift/Sources/ADBuilder/LinkResolver.swift
// Global link resolver — maps every known external URL pattern to its corpus
// storage key, and packages the rules into the `linkResolver` callback shape the
// Markdown→HTML fallback consumes (and the `classify` form the link audit uses).
//
// Native port of `src/lib/link-resolver.js`. URL→key rules live in one place so
// adding a source is one entry, not a per-adapter edit.
//
// Parity note: the JS uses WHATWG `new URL()`; this uses Foundation
// `URLComponents`. For the apple-docs corpus (standard https URLs, ASCII paths)
// the two agree; `host` is lowercased here to match WHATWG, and `percentEncodedPath`
// is used so the rule patterns see the same (encoded) pathname JS does.

import Foundation

public struct LinkResolver: Sendable {
    let knownKeys: Set<String>?
    let swiftOrgPaths: Set<String>?
    let docsBase: String
    let base: URL?

    /// - Parameters:
    ///   - knownKeys: corpus keys that exist; when supplied every pattern match is
    ///     verified before internalization (else strict pattern rules are trusted).
    ///   - swiftOrgPaths: curated swift.org paths to internalize to `/docs/swift-org/<path>/`.
    ///   - sourceURL: absolute URL of the page being parsed, to resolve relative links.
    ///   - docsBase: internal route prefix (default `/docs`).
    public init(
        knownKeys: Set<String>? = nil, swiftOrgPaths: Set<String>? = nil,
        sourceURL: String? = nil, docsBase: String = "/docs"
    ) {
        self.knownKeys = knownKeys
        self.swiftOrgPaths = swiftOrgPaths
        self.docsBase = docsBase
        self.base = sourceURL.flatMap { URL(string: $0) }
    }

    // MARK: - The resolver callback (htmlToMarkdown linkResolver opt)

    /// Returns the rewritten in-corpus `/docs/<key>/` href, the absolute external
    /// URL, or the original href unchanged. Port of `createLinkResolver`'s callback.
    public func resolve(_ rawHref: String) -> String {
        if rawHref.isEmpty { return rawHref }
        // Bail on schemes we never rewrite.
        let lower = rawHref.lowercased()
        if lower.hasPrefix("mailto:") || lower.hasPrefix("tel:") || lower.hasPrefix("javascript:")
            || lower.hasPrefix("data:") || rawHref.hasPrefix("#")
        {
            return rawHref
        }

        guard let abs = Self.absolute(rawHref, relativeTo: base) else { return rawHref }

        // 1. Already an internal /docs/<key>/ route — leave alone.
        if Self.sameOrigin(abs, base) && abs.percentEncodedPath.hasPrefix("\(docsBase)/") {
            return rawHref
        }

        // 2. Internalize against the structured-pattern rules (verify with knownKeys).
        if let candidate = Self.mapURLToKey(abs),
            knownKeys == nil || knownKeys!.contains(candidate)
        {
            return "\(docsBase)/\(candidate)/\(Self.fragment(abs))"
        }

        // 3. swift.org generic-path opt-in: only when the path is curated.
        if let paths = swiftOrgPaths, Self.isSwiftOrg(abs.host) {
            let path = Self.trimSlashes(abs.percentEncodedPath)
            for variant in [path, "\(path).html"] where paths.contains(variant) {
                return "\(docsBase)/swift-org/\(variant)/\(Self.fragment(abs))"
            }
        }

        // 4. Otherwise the absolute URL (at least the link works as external content).
        return Self.whatwgString(abs) ?? rawHref
    }

    /// WHATWG `URL.toString()` serialization detail Foundation omits: a URL with
    /// a host and an EMPTY path prints with `/` (`https://example.com` →
    /// `https://example.com/`). The JS resolver returns `url.toString()`, so the
    /// external-fallback bytes must match.
    static func whatwgString(_ comps: URLComponents) -> String? {
        var copy = comps
        if copy.percentEncodedPath.isEmpty && copy.host != nil {
            copy.percentEncodedPath = "/"
        }
        return copy.string
    }

    // MARK: - Pure URL→key mapping

    /// Try every URL→key rule in order; first non-null match wins. Pure, no I/O.
    public static func mapUrlToKey(_ url: String) -> String? {
        guard !url.isEmpty, let comps = absolute(url, relativeTo: nil) else { return nil }
        return mapURLToKey(comps)
    }

    static func mapURLToKey(_ comps: URLComponents) -> String? {
        let host = comps.host?.lowercased() ?? ""
        let path = comps.percentEncodedPath

        // swift.org redirect aliases win over the generic rules.
        if isSwiftOrg(host) {
            let p = trimSlashes(path)
            if let redirect = swiftOrgRedirects[p] { return redirect }
        }

        // developer.apple.com/documentation/<rest> → <rest> (lowercased)
        if host == "developer.apple.com", path.hasPrefix("/documentation/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/documentation/".count)))
            return rest.isEmpty ? nil : rest.lowercased()
        }
        // developer.apple.com/design/<rest> → design/<rest> (lowercased)
        if host == "developer.apple.com", path.hasPrefix("/design/") {
            let rest = trimTrailingSlashes(dropLeadingSlash(path))
            return rest.isEmpty ? nil : rest.lowercased()
        }
        // developer.apple.com/library/archive/<rest> → apple-archive/<rest> (original case)
        if host == "developer.apple.com", path.hasPrefix("/library/archive/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/library/archive/".count)))
            if rest.isEmpty { return nil }
            if let m = matchArchiveHtml(rest) {
                let parent = m.dir.split(separator: "/").last.map(String.init) ?? ""
                if m.base.lowercased() == "index" || m.base.lowercased() == parent.lowercased() {
                    return "apple-archive/\(m.dir)"
                }
                return "apple-archive/\(m.dir)/\(m.base).\(m.ext.lowercased())"
            }
            return "apple-archive/\(rest)"
        }
        // developer.apple.com/videos/play/wwdcYYYY/ID → wwdc/wwdcYYYY-ID
        if host == "developer.apple.com", let wwdc = matchWwdc(path) {
            return "wwdc/\(wwdc.year)-\(wwdc.id)"
        }
        // docs.swift.org/swift-book/<rest> → swift-book/<rest>
        if host == "docs.swift.org", path.hasPrefix("/swift-book/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/swift-book/".count)))
            return rest.isEmpty ? "swift-book" : "swift-book/\(rest)"
        }
        // docs.swift.org/compiler/<rest> → swift-compiler/<rest>
        if host == "docs.swift.org", path.hasPrefix("/compiler/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/compiler/".count)))
            return rest.isEmpty ? "swift-compiler" : "swift-compiler/\(rest)"
        }
        // docs.swift.org/swiftpm/<rest> → swift-package-manager/<rest>
        if host == "docs.swift.org", path.hasPrefix("/swiftpm/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/swiftpm/".count)))
            return rest.isEmpty ? "swift-package-manager" : "swift-package-manager/\(rest)"
        }
        // swift.org/migration/<rest> → swift-migration-guide/<rest>
        if isSwiftOrg(host), path.hasPrefix("/migration/") {
            let rest = trimTrailingSlashes(String(path.dropFirst("/migration/".count)))
            return rest.isEmpty ? "swift-migration-guide" : "swift-migration-guide/\(rest)"
        }
        // swift.org/swift-evolution/proposals/NNNN-… → swift-evolution/NNNN-…
        if isSwiftOrg(host), path.hasPrefix("/swift-evolution/proposals/"),
            startsWithProposalNumber(String(path.dropFirst("/swift-evolution/proposals/".count)))
        {
            return matchProposalId(path).map { "swift-evolution/\($0)" }
        }
        // github.com/(apple|swiftlang)/swift-evolution/(blob|tree)/… → swift-evolution/NNNN-…
        if host == "github.com", isGithubSwiftEvolution(path) {
            return matchProposalId(path).map { "swift-evolution/\($0)" }
        }
        return nil
    }

    // MARK: - Link classification (for the build's link audit)

    public enum LinkClass: Equatable, Sendable {
        case fragment
        case internalOk(key: String)
        case internalBroken(key: String)
        case externalResolvable(key: String, normalized: String)
        case external(normalized: String)
        case relativeBroken(normalized: String?)
    }

    /// Classify a single URL for audit purposes. Port of `classifyLink`.
    public static func classify(_ url: String, knownKeys: Set<String>, docsBase: String = "/docs")
        -> LinkClass
    {
        if url.isEmpty { return .relativeBroken(normalized: nil) }
        if url.hasPrefix("#") { return .fragment }

        let lower = url.lowercased()
        if lower.hasPrefix("mailto:") || lower.hasPrefix("tel:") || lower.hasPrefix("javascript:")
            || lower.hasPrefix("data:")
        {
            return .external(normalized: url)
        }

        // Internal /docs/<key>/ — verify the key resolves.
        if url.hasPrefix("\(docsBase)/") {
            let key = docsKey(url, docsBase: docsBase)
            if let key, knownKeys.contains(key) { return .internalOk(key: key) }
            return .internalBroken(key: key ?? url)
        }

        // Any other host-relative path — flag as broken.
        if url.hasPrefix("/") { return .relativeBroken(normalized: url) }

        // Absolute URL: try to internalize, else external.
        guard let comps = absolute(url, relativeTo: nil) else {
            return .relativeBroken(normalized: url)
        }
        if let candidate = mapURLToKey(comps), knownKeys.contains(candidate) {
            return .externalResolvable(key: candidate, normalized: url)
        }
        return .external(normalized: url)
    }

    // MARK: - swift.org redirects + host check

    static let swiftOrgRedirects: [String: String] = [
        "documentation/concurrency": "swift-migration-guide/documentation/migrationguide",
        "documentation/package-manager": "swift-package-manager/documentation/packagemanagerdocs",
        "documentation/tspl": "swift-book/The-Swift-Programming-Language",
    ]

    static func isSwiftOrg(_ host: String?) -> Bool {
        host == "swift.org" || host == "www.swift.org"
    }

    // MARK: - URL helpers (WHATWG-equivalent accessors over URLComponents)

    /// Resolve `rawHref` (optionally against `base`) into an ABSOLUTE URL's
    /// components, or nil when it is not absolute (no scheme/host) — mirroring
    /// `new URL()` throwing on a relative ref with no base.
    static func absolute(_ rawHref: String, relativeTo base: URL?) -> URLComponents? {
        let resolved: URL?
        if let base { resolved = URL(string: rawHref, relativeTo: base)?.absoluteURL } else {
            resolved = URL(string: rawHref)
        }
        guard let resolved,
            let comps = URLComponents(url: resolved, resolvingAgainstBaseURL: true),
            comps.scheme != nil, comps.host != nil
        else { return nil }
        return comps
    }

    /// `#fragment` (encoded, with leading `#`) or `""` — JS `u.hash`.
    static func fragment(_ comps: URLComponents) -> String {
        comps.percentEncodedFragment.map { "#\($0)" } ?? ""
    }

    static func sameOrigin(_ a: URLComponents, _ base: URL?) -> Bool {
        guard let base, let b = URLComponents(url: base, resolvingAgainstBaseURL: true) else {
            return true  // no base → JS `base?.origin ?? abs.origin` makes this always true
        }
        return a.scheme?.lowercased() == b.scheme?.lowercased()
            && a.host?.lowercased() == b.host?.lowercased() && a.port == b.port
    }

    // MARK: - String trimming (regex-free)

    static func trimTrailingSlashes(_ s: String) -> String {
        var end = s.endIndex
        while end > s.startIndex, s[s.index(before: end)] == "/" { end = s.index(before: end) }
        return String(s[s.startIndex ..< end])
    }

    static func dropLeadingSlash(_ s: String) -> String {
        s.hasPrefix("/") ? String(s.dropFirst()) : s
    }

    static func trimSlashes(_ s: String) -> String {
        var start = s.startIndex
        while start < s.endIndex, s[start] == "/" { start = s.index(after: start) }
        return trimTrailingSlashes(String(s[start...]))
    }

    // MARK: - Pattern matchers (regex-free, parity with the JS regexes)

    /// `/^\/videos\/play\/(wwdc\d{4})\/(\d+)\/?$/`
    static func matchWwdc(_ path: String) -> (year: String, id: String)? {
        let parts = path.split(separator: "/", omittingEmptySubsequences: true)
        guard parts.count == 4, parts[0] == "videos", parts[1] == "play" else { return nil }
        let year = parts[2], id = parts[3]
        guard year.hasPrefix("wwdc"), year.count == 8,
            year.dropFirst(4).allSatisfy(\.isNumber),
            !id.isEmpty, id.allSatisfy(\.isNumber)
        else { return nil }
        return (String(year), String(id))
    }

    /// The `\d{4}-` prefix test for `/swift-evolution/proposals/<rest>`.
    static func startsWithProposalNumber(_ rest: String) -> Bool {
        let four = rest.prefix(4)
        return four.count == 4 && four.allSatisfy(\.isNumber) && rest.dropFirst(4).first == "-"
    }

    /// `\/proposals\/(\d{4}-[^/]+?)(?:\.md|\.html)?\/?$` → the `NNNN-name` id.
    static func matchProposalId(_ path: String) -> String? {
        guard let r = path.range(of: "/proposals/") else { return nil }
        var rest = trimTrailingSlashes(String(path[r.upperBound...]))
        if rest.hasSuffix(".md") { rest = String(rest.dropLast(3)) } else if rest.hasSuffix(".html")
        {
            rest = String(rest.dropLast(5))
        }
        guard !rest.contains("/"), startsWithProposalNumber(rest) else { return nil }
        return rest
    }

    /// `/^\/(?:apple|swiftlang)\/swift-evolution\/(?:blob|tree)\//`
    static func isGithubSwiftEvolution(_ path: String) -> Bool {
        for owner in ["apple", "swiftlang"] {
            for kind in ["blob", "tree"] where path.hasPrefix("/\(owner)/swift-evolution/\(kind)/") {
                return true
            }
        }
        return false
    }

    /// `/^(.*)\/([^/]+)\.(html|htm)$/i` → (dir, base, ext) where ext is original case.
    static func matchArchiveHtml(_ s: String) -> (dir: String, base: String, ext: String)? {
        guard let dot = s.range(of: ".", options: .backwards) else { return nil }
        let ext = String(s[dot.upperBound...])
        let lowerExt = ext.lowercased()
        guard lowerExt == "html" || lowerExt == "htm" else { return nil }
        let beforeExt = String(s[..<dot.lowerBound])
        guard let slash = beforeExt.range(of: "/", options: .backwards) else { return nil }
        let base = String(beforeExt[slash.upperBound...])
        guard !base.isEmpty else { return nil }
        return (String(beforeExt[..<slash.lowerBound]), base, ext)
    }

    /// The corpus key from a `<docsBase>/<key>/?(#|?…)` URL — `^([^?#]+?)\/?(?:[?#].*)?$`
    /// then percent-decoded (JS `decodeURIComponent`).
    static func docsKey(_ url: String, docsBase: String) -> String? {
        let afterBase = String(url.dropFirst(docsBase.count + 1))
        // Cut at the first ? or #.
        let body = afterBase.prefix { $0 != "?" && $0 != "#" }
        let key = trimTrailingSlashes(String(body))
        guard !key.isEmpty else { return nil }
        return key.removingPercentEncoding ?? key
    }
}
